experimental/packages/opentelemetry-instrumentation-fetch/test/fetch.test.ts

```diff
@@ -239,36 +238,25 @@ describe('fetch', () => {
     );
 
     rootSpan = webTracerWithZone.startSpan('root');
-    api.context.with(api.trace.setSpan(api.context.active(), rootSpan), () => {
+    await api.context.with(api.trace.setSpan(api.context.active(), rootSpan), async () => {
       fakeNow = 0;
-      void getData(fileUrl, method)
-        .then(
-          response => {
-            // this is a bit tricky as the only way to get all request headers from
-            // fetch is to use json()
-            return response.json().then(
-              json => {
-                lastResponse = json;
-                const headers: { [key: string]: string } = {};
-                Object.keys(lastResponse.headers).forEach(key => {
-                  headers[key.toLowerCase()] = lastResponse.headers[key];
-                });
-                lastResponse.headers = headers;
-              },
-              () => {
-                lastResponse = undefined;
-              }
-            );
-          },
-          () => {
-            lastResponse = undefined;
-          }
-        )
-        .then(sinon.clock.runAllAsync)
-        .then(() => {
-          done();
```

Comment:
IIUC, this `done` should handle the asynchrony of this procedure. Would you mind elaborating on how this change addresses the problem?

Reply:
I am not completely sure. I thought at first that maybe calling `done` before the end of execution inside context caused the issue, but I have found a minimal patch that fixes the issue and disproves the context theory.

This patch is still flaky:
```diff
diff --git a/experimental/packages/opentelemetry-instrumentation-fetch/test/fetch.test.ts b/experimental/packages/opentelemetry-instrumentation-fetch/test/fetch.test.ts
index a38dbd8b..1e554c96 100644
--- a/experimental/packages/opentelemetry-instrumentation-fetch/test/fetch.test.ts
+++ b/experimental/packages/opentelemetry-instrumentation-fetch/test/fetch.test.ts
@@ -241,7 +241,8 @@ describe('fetch', () => {
     rootSpan = webTracerWithZone.startSpan('root');
     api.context.with(api.trace.setSpan(api.context.active(), rootSpan), () => {
       fakeNow = 0;
-      void getData(fileUrl, method)
+      const p = getData(fileUrl, method);
+      Promise.resolve().then(() => p)
         .then(
           response => {
             // this is a bit tricky as the only way to get all request headers from
```

This one does not appear to be flaky any more (just `async` keyword is added):
```diff
diff --git a/experimental/packages/opentelemetry-instrumentation-fetch/test/fetch.test.ts b/experimental/packages/opentelemetry-instrumentation-fetch/test/fetch.test.ts
index a38dbd8b..41338be1 100644
--- a/experimental/packages/opentelemetry-instrumentation-fetch/test/fetch.test.ts
+++ b/experimental/packages/opentelemetry-instrumentation-fetch/test/fetch.test.ts
@@ -241,7 +241,8 @@ describe('fetch', () => {
     rootSpan = webTracerWithZone.startSpan('root');
     api.context.with(api.trace.setSpan(api.context.active(), rootSpan), () => {
       fakeNow = 0;
-      void getData(fileUrl, method)
+      const p = getData(fileUrl, method);
+      Promise.resolve().then(async () => p)
         .then(
           response => {
             // this is a bit tricky as the only way to get all request headers from
```

Reply:
That sounds a bit like a race-condition in the task loop.
Using timers and scheduling immediately resolving promises(happens a lot in mocking) is tricky anyways, but I think it's worth first finding out a reason it's flaky.

Reply:
I think, it's this issue in Zone.js: https://github.com/angular/angular/issues/44446

Reply:
I think, I figured out the reason it is failing. The span ends only after the instrumentation reads the whole body: 
https://github.com/open-telemetry/opentelemetry-js/blob/610808d/experimental/packages/opentelemetry-instrumentation-fetch/src/fetch.ts#L350-L352

But this is not tied to the fetch promise resolution in any way:
https://github.com/open-telemetry/opentelemetry-js/blob/610808d/experimental/packages/opentelemetry-instrumentation-fetch/src/fetch.ts#L367

So, we've got the response, the promise chain starts at the same time the reading starts. If reading is slow, `done` will be called before the body is read. Which means there's no exported span yet.

Adding `async` seems to change the order in which the microtasks are scheduled for execution, and that solves the issue. It's not a very obvious, though.

Reply:
IMO it's not really a fix. You don't fix a race condition by making one of the racers faster or slower. It sounds like it is still a race, and one that depends on the underlying internals, cpu speeds, IO speeds, et cetera, which makes it difficult to guarantee it won't come back up. I would prefer to force the execution order if possible

Reply:
I agree, it's a workaround which seems to be working with the current internals. I could've worded it better.
Mostly, I just wanted to share my findings.

You don't promote a welcoming community by "you don'ting" at people. 😛

Reply:
Sorry if I caused offense it wasn't my intention. I meant to ask if you think there is a way this can be modified to force execution order rather than applying a workaround so we can be sure it will continue to work in the future if zone.js is updated for example.

Reply:
We can wait for `_endSpan` to be called, like this: https://github.com/open-telemetry/opentelemetry-js/pull/2969/commits/b4777abdd9de5b6810a2a1d9c526d099cb6f9864